### PR TITLE
Use cmake_dependent option for BLT options

### PR DESCRIPTION
**Pull request by gzagaris**
# Summary

This PR proposes using `cmake_dependent_option` for options that are defined by BLT. This avoids shadowing of the BLT options by options redefined in RAJA and in the cases where RAJA is used a sub-module within another BLT project, it provides the ability to disable tests/examples etc. at a more fine granularity.

### CODE REVIEW

**Comment by gzagaris** [comment]
@rhornung67 , @davidbeckingsale, at your convenience, could you take a quick look at this?

**Comment by rhornung67** [comment]
LGTM

**Review by rhornung67 — CHANGES_REQUESTED** [review]
@gzagaris please add the same logic for the RAJA exercises.

**Comment by rhornung67** [comment]
@davidbeckingsale I'm good with this if building the exercises follows the same logic.

**Comment by davidbeckingsale** [comment]
ENABLE_EXERCISES doesn't shadow a BLT option though. I think that can be done later when we prefix all options.

**Review by davidbeckingsale — APPROVED** [review]

File: docs/sphinx/user_guide/config_options.rst [hunk 1]
```diff
@@ -89,8 +89,8 @@ and their default settings:
       ======================   ======================
       Variable                 Default
       ======================   ======================
-      ENABLE_TESTS             On 
-      ENABLE_EXAMPLES          On 
+      RAJA_ENABLE_TESTS        On 
+      RAJA_ENABLE_EXAMPLES     On
```
Comment (MrBurmark) [inline docs/sphinx/user_guide/config_options.rst:93]: Should we add RAJA_ENABLE_BENCHMARKS to this?

Comment (rhornung67) [reply docs/sphinx/user_guide/config_options.rst:93]: I don't think we need that. See David's comment above.

**Review by rhornung67 — APPROVED** [review]

**Comment by gzagaris** [comment]
> ENABLE_EXERCISES doesn't shadow a BLT option though. I think that can be done later when we prefix all options.

That's right, it is only defined in RAJA. But, it would be good to be consistent though. I would suggest that all options have a `RAJA_` prefix for the same reasons. I'll go ahead and the `RAJA_` prefix to `ENABLE_EXERCISES`, but this does not need to be a `cmake_depende_option` since it is not defined by BLT.

File: docs/sphinx/user_guide/config_options.rst (hunk 1 above)
Comment (gzagaris) [reply docs/sphinx/user_guide/config_options.rst:93]: ENABLE_BENCHMARKS is defined by BLT, so that is needed -- I'll make that change too.

**Comment by rhornung67** [comment]
> > ENABLE_EXERCISES doesn't shadow a BLT option though. I think that can be done later when we prefix all options.
> 
> That's right, it is only defined in RAJA. But, it would be good to be consistent though. I would suggest that all options have a `RAJA_` prefix for the same reasons. I'll go ahead and the `RAJA_` prefix to `ENABLE_EXERCISES`, but this does not need to be a `cmake_depende_option` since it is not defined by BLT.

We plan to clean up our cmake options to add the 'RAJA_' prefix for those that are ours, avoid redundancy with BLT, etc. If you want to make additional modifications, please feel free, but don't feel obligated since it will be addressed soon by us.

File: docs/sphinx/user_guide/config_options.rst (hunk 1 above)
Comment (gzagaris) [reply docs/sphinx/user_guide/config_options.rst:93]: Done.

**Comment by gzagaris** [comment]
> > > ENABLE_EXERCISES doesn't shadow a BLT option though. I think that can be done later when we prefix all options.
> > 
> > 
> > That's right, it is only defined in RAJA. But, it would be good to be consistent though. I would suggest that all options have a `RAJA_` prefix for the same reasons. I'll go ahead and the `RAJA_` prefix to `ENABLE_EXERCISES`, but this does not need to be a `cmake_depende_option` since it is not defined by BLT.
> 
> We plan to clean up our cmake options to add the 'RAJA_' prefix for those that are ours, avoid redundancy with BLT, etc. If you want to make additional modifications, please feel free, but don't feel obligated since it will be addressed soon by us.

Sounds good -- Thanks @rhornung67!

I made the changes you requested since those were quick. I am glad to hear that you are moving towards cleaning up all the options and add the `RAJA_` prefix. This would be very helpful and I look forward to checking out a fresh version with these changes.

I also made some changes in Umpire and camp along the same lines (I got PRs up on those as well), but, I only touched the things that were pertaining to my immediate needs. I hope that Umpire and camp will follow this lead and clean up the respective options as well.

**Comment by gzagaris** [comment]
Thanks @davidbeckingsale  and @MrBurmark for your feedback and reviewing this PR as well.

**Review by MrBurmark — APPROVED** [review]

**Comment by trws** [comment]
Sounds like a good plan.  I'm good with this change here, and I'll head over to camp to look that one over.  Likely will either take it or do something to get similar behavior since we're in a non-BLT world over there still.

**Review by rhornung67 — APPROVED** [review]

**Comment by rhornung67** [comment]
LGTM

**Comment by rhornung67** [comment]
LGTM

**Comment by rhornung67** [comment]
@davidbeckingsale are you good with this PR? If so, wanna pull the trigger on the Gitlab CI?

**Comment by davidbeckingsale** [comment]
LGTM

**Comment by rhornung67** [comment]
Does anyone see why travis is failing on this? Also, it is only running 15 tests. So wither the CMake logic is not right, or we need to change the options passed to CMake in the travis YAML file.

**Comment by davidbeckingsale** [comment]
> Does anyone see why travis is failing on this? Also, it is only running 15 tests. So wither the CMake logic is not right, or we need to change the options passed to CMake in the travis YAML file.

I think it's because the dependent option is created before BLT is loaded, and BLT defines e.g. the ENABLE_TESTS option. So either the dependent options should come after BLT is loaded, or we need to explicitly set ENABLE_TESTS etc. in our testing and CI. I think the former is more in line with the current behavior and would be my preference.

**Comment by rhornung67** [comment]
@davidbeckingsale thanks. I'll play around with this and push changes based on your comment if that causes all tests to run.

**Comment by rhornung67** [comment]
@gzagaris I pulled your branch into our repo and made a new PR with a modification to make this work as intended. Please take a look here: https://github.com/LLNL/RAJA/pull/991

I can't add you as a reviewer to that. Please let us know if you are OK with my change.

**Comment by gzagaris** [comment]
> @gzagaris I pulled your branch into our repo and made a new PR with a modification to make this work as intended. Please take a look here: #991
> 
> I can't add you as a reviewer to that. Please let us know if you are OK with my change.

Thanks, that works for me.

**Comment by rhornung67** [comment]
@gzagaris thanks for taking a look. We'll get this is as soon as we can. We're struggling with Gitlab CI throughput. We're shooting for a RAJA release in a few weeks and this will be in there.

**Comment by gzagaris** [comment]
> @gzagaris thanks for taking a look. We'll get this is as soon as we can. We're struggling with Gitlab CI throughput. We're shooting for a RAJA release in a few weeks and this will be in there.

Thanks @rhornung67 -- I appreciate all the help!

**Pull request merged**